Stream :gen_statem

hpopp opened the pull request:
@cstar Took a crack at the `:gen_statem` implementation myself. Was surprisingly easy to implement with a Registry.

Each stream is spawned with a reference to the original `Kadabra.Connection` and it's ssl socket. I'm guessing a restarted socket will need to be propagated across all of the active streams?

Also worth considering.. this would bump minimum requirements up to Elixir 1.4 and OTP 19.

hpopp committed:
feat: streams now use :gen_statem

hpopp committed:
travis: updated to proper version requirements

hpopp committed:
travis: bumped elixir version to 1.4.2

hpopp committed:
fix: travis config

cstar commented:
Excellent, giving it a shot right now. And a good reason to upgrade to Elixir 1.4.2 :)
Thanks

hpopp committed:
test: added image file GET test

hpopp committed:
Merge branch 'master' into gen_statem

hpopp committed:
refactor: various cleanup + dialyzer

hpopp committed:
travis: otp 19/v1.4.5 runner

hpopp committed:
travis: fixed otp version -> 19.1

hpopp committed:
travis: otp 19.2 runner

hpopp committed:
feat: initial push promise implementation

TODO:
- Handle headers in promise payload
- Tests

hpopp committed:
feat: push_promise parses headers

Also:
- fixed proper stream_id parsing

hpopp committed:
refactor: custom frame structs WIP

hpopp committed:
chore: delete duplicate file

hpopp committed:
refactor: frame-specific modules

Part of a general refactor to move frame logic into their respective
submodules.

hpopp committed:
feat: new SETTINGS/WINDOW_UPDATE WIP

hpopp committed:
refactor: flags, etc

- Flags now consolidatedi into Kadabra.Flags
- Kadabra.Connection is now a struct

hpopp committed:
doc: various typespecs

hpopp committed:
refactor: encodable protocol

hpopp committed:
chore: cleaned up compile warnings

hpopp committed:
test: test for push promise

hpopp committed:
test: added test for auto reconnect
